mosviz/controls/slit_controller.py

```diff
@@ -0,0 +1,204 @@
+import numpy as np
+
+from glue.core import HubListener
+
+from regions import RectangleSkyRegion, RectanglePixelRegion, PixCoord
+
+from matplotlib.patches import Rectangle
+
+from astropy.coordinates import Angle, SkyCoord
+from astropy.wcs.utils import proj_plane_pixel_area
+from astropy import units as u
+
+from ..controls.slit_selection_ui import SlitSelectionUI
+
+
+class SlitController(HubListener):
+    """
+    Controller that constructs and stores
+    a rectangular slit.
+    """
+
+    def __init__(self, mosviz_viewer=None):
+        super(SlitController, self).__init__()
+        self.mosviz_viewer = mosviz_viewer
+
+        self._slit = None  # `region.RectangleSkyRegion` object
+        self._pix_slit = None  # `region.RectanglePixelRegion` object
+        self._patch = None  # `matplotlib.patches.Rectangle` patch
+
+    @property
+    def is_active(self):
+        return self._patch is not None
+
+    @property
+    def patch(self):
+        return self._patch
+
+    @property
+    def x(self):
+        """Center x position of slit"""
+        if self.is_active:
+            return self.patch.get_x() + self.patch.get_width() / 2.
+
+    @property
+    def y(self):
+        """Center y position of slit"""
+        if self.is_active:
+            return self.patch.get_y() + self.patch.get_height() / 2.
+
+    @property
+    def width(self):
+        if self.is_active:
+            return self.patch.get_width()
+
+    @property
+    def length(self):
+        if self.is_active:
+            return self.patch.get_height()
+
+    @property
+    def dx(self):
+        """Width of slit"""
+        return self.width
+
+    @property
+    def dy(self):
+        """Length of slit"""
+        return self.length
+
+    @property
+    def x_bounds(self):
+        """x axis max and min pixel values"""
+        xp = self.x
+        dx = self.dx
+        x_min = xp - dx / 2.
+        x_max = xp + dx / 2.
+        return [x_min, x_max]
+
+    @property
+    def y_bounds(self):
+        """y axis max and min pixel values"""
+        yp = self.y
+        dy = self.dy
+        y_min = yp - dy / 2.
+        y_max = yp + dy / 2.
+        return [y_min, y_max]
+
+    def construct_simple_rectangle(self, x=None, y=None, width=None, length=None):
+        """
+        Simple matplotlib rectangle patch.
+
+        Parameters
+        ----------
+        x, y : float
+            Center (x, y) of slit in pix.
+        width, length : float
+            width and length of slit in pix.
+        Returns
+        -------
+        patch : `matplotlib.patches.Rectangle`
+        """
+        self.destruct()
+
+        self._slit = None
+        self._pix_slit = None
+        self._patch = Rectangle((x - width / 2, y - length / 2),
+                                width=width, height=length,
+                                edgecolor='red', facecolor='none')
+        return self._patch
+
+    def construct_pix_region(self, x, y, width, length):
```

Comment:
I'm confused about why `construct_pix_region` and `construct_sky_region` exist as separate functions that only set the `_pix_slit`/`_slit` and not the other.

One of the core ideas behind regions is that, armed with a WCS, you can convert back and forth between pixel and sky regions (https://astropy-regions.readthedocs.io/en/latest/shapes.html#transformations) - so they shouldn't be "independent" representations - giving one + a WCS you should be able to produce both representations and then only plot the sky one when in sky coordinates and the other in pixel coordinates?

Reply:
I changed the definition of the slit controller and slit:
slit is now a subclass of the `_MOSVizSlit` class. It must implements all abstract attributes and functions.
Slit controller basically has a list of  `_MOSVizSlit` items and is responsible for communicating with the rest of the codebase.